compiler/p/codegen/OMRLoadStoreHandler.cpp

```diff
@@ -0,0 +1,393 @@
+    }
+
+void OMR::Power::LoadStoreHandlerImpl::generateLoadSequence(TR::CodeGenerator *cg, TR::Register *trgReg, TR::Node *node, TR::MemoryReference *memRef, TR::InstOpCode::Mnemonic loadOp)
+    {
+    generateTrg1MemInstruction(cg, loadOp, node, trgReg, memRef);
+    cg->insertPrefetchIfNecessary(node, trgReg);
+
+#ifdef J9_PROJECT_SPECIFIC
+    if (node->getSymbol()->isSyncVolatile() && cg->comp()->target().isSMP())
+        TR::TreeEvaluator::postSyncConditions(node, cg, trgReg, memRef, cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P7) ? TR::InstOpCode::lwsync : TR::InstOpCode::isync);
+#endif
+    }
+
+void OMR::Power::LoadStoreHandlerImpl::generatePairedLoadSequence(TR::CodeGenerator *cg, TR::Register *trgReg, TR::Node *node, TR::MemoryReference *memRef)
+    {
+    #ifdef J9_PROJECT_SPECIFIC
+    // Since non-volatiles are implemented as two separate loads, we must use a special sequence to perform the load in
+    // a single instruction even when SMP is disabled.
+    if (node->getSymbol()->isSyncVolatile())
+        {
+        if (node->getSymbolReference()->isUnresolved())
+            {
+            TR::SymbolReference *vrlSymRef = cg->comp()->getSymRefTab()->findOrCreateVolatileReadLongSymbolRef(cg->comp()->getMethodSymbol());
+
+            memRef->getUnresolvedSnippet()->setIs32BitLong();
+
+            TR::RegisterDependencyConditions *deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(4, 4, cg->trMemory());
+            TR::addDependency(deps, trgReg->getHighOrder(), TR::RealRegister::gr3, TR_GPR, cg);
+            TR::addDependency(deps, trgReg->getLowOrder(), TR::RealRegister::gr4, TR_GPR, cg);
+            TR::addDependency(deps, NULL, TR::RealRegister::gr11, TR_GPR, cg);
+
+            TR::addDependency(deps, memRef->getBaseRegister(), TR::RealRegister::NoReg, TR_GPR, cg);
+            deps->getPreConditions()->getRegisterDependency(3)->setExcludeGPR0();
+            deps->getPostConditions()->getRegisterDependency(3)->setExcludeGPR0();
+
+            generateTrg1MemInstruction(cg, TR::InstOpCode::addi2, node, trgReg->getHighOrder(), memRef);
+            generateDepImmSymInstruction(
+                cg,
+                TR::InstOpCode::bl,
+                node,
+                reinterpret_cast<uintptr_t>(vrlSymRef->getSymbol()->castToMethodSymbol()->getMethodAddress()),
+                deps,
+                vrlSymRef
+            );
+
+            cg->machine()->setLinkRegisterKilled(true);
+            }
+        else if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P8) && cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_PPC_HAS_VSX))
+            {
+            TR::Register *tempHiReg = cg->allocateRegister(TR_FPR);
+            TR::Register *tempLoReg = cg->allocateRegister(TR_FPR);
+
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lfd, node, tempHiReg, memRef);
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::xxspltw, node, tempLoReg, tempHiReg, 1);
```

Comment:
you should choose to splat word[0], instead of word[1]. then, mfvsrwz existing FPR to low order, and the splat one to high order.  as it is, the low word is duplicated into the pair.

unfortunately, we can only test this out in 32bit IBM JDK.

Reply:
Fixed.